synapse/util/caches/lrucache.py

```diff
@@ -82,19 +93,241 @@ def enumerate_leaves(node, depth):
                 yield m
 
 
+P = TypeVar("P")
+LN = TypeVar("LN", bound="_ListNode")
+
+
+class _ListNode(Generic[P]):
+    """A node in a circular doubly linked list, with an (optional) reference to
+    a cache entry.
+
+    The reference should only be `None` for the root node or if the node has
+    been removed from the list.
+    """
+
+    # A lock to protect mutating the list prev/next pointers.
+    _LOCK = threading.Lock()
+
+    # We don't use attrs here as in py3.6 you can't have `attr.s(slots=True)`
+    # and inherit from `Generic` for some reason
+    __slots__ = [
+        "cache_entry",
+        "prev_node",
+        "next_node",
+    ]
+
+    def __init__(self, cache_entry: Optional[P] = None) -> None:
+        self.cache_entry = cache_entry
+        self.prev_node: Optional[_ListNode[P]] = None
+        self.next_node: Optional[_ListNode[P]] = None
+
+    @classmethod
+    def create_root_node(cls: Type["_ListNode[P]"]) -> "_ListNode[P]":
+        root = cls()
+        root.prev_node = root
+        root.next_node = root
+        return root
+
+    @classmethod
+    def insert_after(
+        cls: Type[LN],
+        cache_entry: P,
+        node: "_ListNode[P]",
+    ) -> LN:
+        """Create a new list node that is placed after the given node."""
+        new_node = cls(cache_entry)
+        with cls._LOCK:
+            new_node._refs_insert_after(node)
+        return new_node
+
+    def remove_from_list(self):
+        """Remove this node from the list."""
+        with self._LOCK:
+            self._refs_remove_node_from_list()
+
+        self.cache_entry = None
+
+    def move_after(self, root: "_ListNode"):
```

Comment:
can we rename `root` to `node` or `other` or `target` or something? It's not limited to being the root node afaict.